src/fb.c

```diff
@@ -1,63 +1,275 @@
 /* SPDX-License-Identifier: MIT */
 
+#include <assert.h>
+
 #include "fb.h"
+#include "smp.h"
+#include "string.h"
 #include "utils.h"
 #include "xnuboot.h"
 
-u32 *fb;
-int fb_s, fb_w, fb_h, fb_d;
-int scale = 1;
+#define FB_DEPTH_FLAG_RETINA 0x10000
+#define FB_DEPTH_MASK        0xff
+
+#define CONSOLE_MAX_ROWS 60
+#define CONSOLE_MAX_COLS 240
+
+#define CONSOLE_MARGIN_LEFT   30
+#define CONSOLE_MARGIN_RIGHT  30
+#define CONSOLE_MARGIN_TOP    50
+#define CONSOLE_MARGIN_BOTTOM 50
+
+fb_t fb;
+
+typedef struct {
+    u32 row;
+    u32 col;
+} console_pos_t;
+
+static struct {
+    u32 *ptr;
+    u32 width;
+    u32 height;
+    u32 x;
+    u32 y;
+
+    u32 enabled;
+} logo = {.enabled = 0};
+
+static struct {
+    u32 initialized;
+
+    struct {
+        u8 *ptr;
+        u32 width;
+        u32 height;
+    } font;
 
-static u32 *logo;
-static int logo_w, logo_h;
+    console_pos_t cursor;
+
+    u32 n_rows;
+    u32 n_cols;
+    u32 row_offset;
+
+    u8 text[CONSOLE_MAX_ROWS][CONSOLE_MAX_COLS];
+} console = {.initialized = 0,
+             .cursor = {.row = 0, .col = 0},
+             .n_rows = CONSOLE_MAX_ROWS,
+             .n_cols = CONSOLE_MAX_COLS,
+             .row_offset = 0,
+             .text = {{0}}};
```

Comment:
In general I keep public interfaces/structs/etc. in .h files with the expectation that these won't change. Internal structure on the other hand are kept inside .c files. These structure are only used internally and I'd like to keep them there.

Reply:
That's fine then, I think the biggest confusing thing is that the closing brace for console is not on it's own line, which visually makes that section look wrong to me. But that's just personal preference (and it might be correct bracing I didn't check elsewhere), so I guess it's okay to leave as is.

Reply:
Yeah, I'd prefer to have the brackets a little bit different as well but this is what `clang-format` did and I just accepted it 😄